File: 3986-MaximumPathScoreInAGrid/3986-MaximumPathScoreInAGrid.cpp
// Last updated: 12/25/2025, 7:09:29 PM
#include <cstring>
#include <iostream>
#include <vector>

using namespace std;

int inf = INT_MAX;
class Solution {
public:
    int maxPathScore(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        k = min(k, 402);

        int dp[m][n][k + 1];

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                for (int l = 0; l <= k; ++l) {
                    dp[i][j][l] = -inf;
                }
            }
        }

        dp[0][0][k] = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                for (int rem = 0; rem <= k; rem++) {

                    for (int i2=0; i2<=1; i2++) {
                        for (int j2=0; j2<=1; j2++) {
                            if (i2 == j2) continue;
                            int i3 = i2 + i;
                            int j3 = j2 + j;
                            if (i3 < m && j3 < n && (grid[i3][j3] == 0 || rem >= 1)) {
                                int new_rem = rem - (grid[i3][j3] == 0 ? 0 : 1);
                                // cout << i3 << " " << j3 << " " << rem << endl;
                                dp[i3][j3][new_rem] = max(dp[i3][j3][new_rem], dp[i][j][rem] + grid[i3][j3]);
                                // cout << "ANS: " << dp[i][j][rem] << " " << dp[i3][j3][new_rem] << endl;
                            }
                        }
                    }

                }
            }
        }
        
        int ans = -1;
        for (int rem=0; rem<=k; rem++) {
            ans = max(ans, dp[m - 1][n - 1][rem]);
        }
        return ans;
    }
};

File: 3986-MaximumPathScoreInAGrid/3986-MaximumPathScoreInAGrid.py
# Last updated: 12/25/2025, 7:09:31 PM
cmax = lambda x, y: x if x > y else y
cmin = lambda x, y: x if x < y else y
class Solution:
    def maxPathScore(self, grid: List[List[int]], k: int) -> int:
        # k = 1000
        # grid = [[random.randint(0, 3) for _ in range(200)] for _ in range(200)]
        
        m = len(grid)
        n = len(grid[0])

        dp = [[defaultdict(lambda: -inf) for _ in range(n)] for _ in range(m)]
        dp[0][0][k] = 0
        
        def getCostCell(cell):
            return 1 if cell > 0 else 0
        
        def getCost(i, j):
            return getCostCell(grid[i][j])

        cost = 0
        for i in range(m):
            for j in range(n):
                for rem in dp[i][j]:
                    if i + 1 < m:
                        cell = grid[i+1][j]
                        cost = getCostCell(cell)
                        if rem >= cost:
                            cost = rem - cost
                            dp[i + 1][j][cost] = cmax(dp[i + 1][j][cost], dp[i][j][rem] + cell)
                    if j + 1 < n:
                        cell = grid[i][j+1]
                        cost = getCostCell(cell)
                        if rem >= cost:
                            cost = rem - cost
                            dp[i][j + 1][cost] = cmax(dp[i][j + 1][cost], dp[i][j][rem] + cell)

        ans = -inf
        for rem in dp[m-1][n-1]:
            ans = max(ans, dp[m-1][n-1][rem])
        return -1 if ans == -inf else ans